core/src/test/scala/org/apache/spark/network/netty/NettyBlockTransferSuite.scala

```diff
@@ -0,0 +1,154 @@
+    to.init(blockManager)
+
+    try {
+      val promise = Promise[ManagedBuffer]()
+
+      to.fetchBlocks(from.hostName, from.port, "1", Array(blockId.toString),
+        new BlockFetchingListener {
+          override def onBlockFetchFailure(blockId: String, exception: Throwable): Unit = {
+            promise.failure(exception)
+          }
+
+          override def onBlockFetchSuccess(blockId: String, data: ManagedBuffer): Unit = {
+            promise.success(data.retain())
+          }
+        })
+
+      Await.ready(promise.future, FiniteDuration(100, TimeUnit.SECONDS))
+      promise.future.value.get.get
+    } finally {
+      from.close()
+      to.close()
+    }
+
+  }
+
+  ignore("simple fetch") {
+    val blockString = "Hello, world!"
+    val blockBuffer = LargeByteBufferHelper.asLargeByteBuffer(blockString.getBytes)
+    val fetched = fetchBlock(blockBuffer)
+
+    IOUtils.toString(fetched.createInputStream()) should equal(blockString)
+  }
+
+
+  def uploadBlock(buf: LargeByteBuffer, rddId: Int, timeout: Long) {
+
+    val fromBlockManager = mock[BlockDataManager]
+    val toBlockManager = mock[BlockDataManager]
+    val blockId = RDDBlockId(rddId, rddId + 1)
+    val blockBuffer = new NioManagedBuffer(buf)
+    val level = StorageLevel.DISK_ONLY //doesn't matter
+
+    val from = new NettyBlockTransferService(conf, securityManager, numCores = 1)
+    from.init(fromBlockManager)
+    val to = new NettyBlockTransferService(conf, securityManager, numCores = 1)
+    to.init(toBlockManager)
+
+    try {
+      val uploadFuture = from.uploadBlock(to.hostName, to.port, "exec-1", blockId, blockBuffer, level)
+      Await.result(uploadFuture, Duration.apply(timeout, TimeUnit.MILLISECONDS))
+      val bufferCaptor = ArgumentCaptor.forClass(classOf[ManagedBuffer])
+      verify(toBlockManager).putBlockData(MockitoMatchers.eq(blockId), bufferCaptor.capture(),
+        MockitoMatchers.eq(level))
+      val putBuffer = bufferCaptor.getValue()
+      logTrace("begin checking buffer equivalence")
+      equivalentBuffers(blockBuffer, putBuffer)
+      logTrace("finished checking buffer equivalence")
+    } finally {
+      from.close()
+      to.close()
+    }
+
+  }
+
+  test("small one-part upload") {
+    val buf = LargeByteBufferHelper.asLargeByteBuffer(Array[Byte](0,1,2,3))
+    uploadBlock(buf, 0, 100)
+  }
+
+  test("small multi-part upload") {
+    val parts = (0 until 5).map{idx =>
+      val arr = Array.tabulate[Byte](100){subIdx => (idx + subIdx).toByte}
+      ByteBuffer.wrap(arr)
+    }.toArray
+    val buf = new WrappedLargeByteBuffer(parts)
+    uploadBlock(buf, 1, 500)
+  }
+
+  //just don't want to kill the test server
+  ignore("giant upload") {
+    // pretty close to max size due to overhead from the rest of the msg
+    val parts = (0 until 2).map{_ => ByteBuffer.allocate(Integer.MAX_VALUE - 200)}.toArray
+    val buf = new WrappedLargeByteBuffer(parts)
+    uploadBlock(buf, 2, 20 * 60 * 1000) // yup, takes this long ...
```

Comment:
this test does pass on my machine, when I run it with a 16 gb heap, but didn't want to make it run on jenkins for now